src/octree/batch_iterator.rs

```diff
@@ -0,0 +1,206 @@
+struct PointStream<'a, F>
+where
+    F: FnMut(PointData) -> Result<()>,
+{
+    position: Vec<Vector3<f64>>,
+    color: Vec<Vector4<u8>>,
+    intensity: Vec<f32>,
+    func: &'a mut F,
+}
+
+impl<'a, F> PointStream<'a, F>
+where
+    F: FnMut(PointData) -> Result<()>,
+{
+    fn new(num_points_per_batch: usize, func: &'a mut F) -> Self {
+        PointStream {
+            position: Vec::with_capacity(num_points_per_batch),
+            color: Vec::with_capacity(num_points_per_batch),
+            intensity: Vec::with_capacity(num_points_per_batch),
+            func,
+        }
+    }
+
+    /// push point in batch
+    fn push_point(&mut self, point: Point) {
+        self.position.push(point.position);
+        self.color.push(Vector4::new(
+            point.color.red,
+            point.color.green,
+            point.color.blue,
+            point.color.alpha,
+        ));
+        if let Some(point_intensity) = point.intensity {
+            self.intensity.push(point_intensity);
+        };
+    }
+
+    /// execute function on batch of points
+    fn callback(&mut self) -> Result<()> {
+        if self.position.is_empty() {
+            return Ok(());
+        }
+
+        let mut layers = FnvHashMap::default();
+        layers.insert(
+            "color".to_string(),
+            LayerData::U8Vec4(self.color.split_off(0)),
+        );
+        if !self.intensity.is_empty() {
+            layers.insert(
+                "intensity".to_string(),
+                LayerData::F32(self.intensity.split_off(0)),
+            );
+        }
+        let point_data = PointData {
+            position: self.position.split_off(0),
+            layers,
+        };
+        (self.func)(point_data)
+    }
+}
+
+/// Iterator on point batches
+pub struct BatchIterator<'a> {
+    _octree: &'a Octree,
+    iterator: Box<Iterator<Item = Point> + 'a>,
+    batch_size: usize,
+}
+
+impl<'a> BatchIterator<'a> {
+    pub fn new(
+        octree: &'a octree::Octree,
+        location: &'a octree::batch_iterator::PointLocation,
+        size: usize,
+    ) -> Self {
+        let iter: Box<Iterator<Item = Point>> = match location {
+            PointLocation::Any() => Box::new(octree.all_points()),
+            PointLocation::Aabb(aabb) => Box::new(octree.points_in_box(aabb)),
+            PointLocation::Frustum(frustum) => Box::new(octree.points_in_frustum(frustum)),
+            PointLocation::OrientedBeam(beam) => Box::new(octree.points_in_oriented_beam(beam)),
+        };
+        BatchIterator {
+            _octree: octree,
+            iterator: iter,
+            batch_size: size,
+        }
+    }
+
+    /// compute a funtion while iterating on a batch of points
+    pub fn try_for_each_batch<F>(&mut self, mut func: F) -> Result<()>
+    where
+        F: FnMut(PointData) -> Result<()>,
+    {
+        let mut point_stream = PointStream::new(self.batch_size, &mut func);
+        'octree_loop: loop {
```

Comment:
Is there any specific reason why you decided not to use
`iter.try_for_each(|point: Point| point_stream.push_point_and_callback(point))?;`
from the original implementation?
This loop seems to be harder to understand...

Reply:
Indeed ..My intention is to spare the call of the method(s) and the checks whether the batch function has to be run.

Reply:
I did a test and i spare already 50 ms for an octree of 200_000 points and batch size of 50_000
Current 343.306956ms
Try for each 394.974788ms

Reply:
Did you rerun the test multiple times for each variant?

Reply:
I just tried the test myself wrapping it inside a loop calling it 100 times each. The new implementation calls the `callback` one time less then the original implementation, which would need to be fixed. This difference in calling the callback actually leads to the small time difference, since printing to screen takes quite a while. I would therefore go with the old implementation.

Reply:
Oh that is true, it is missing in line 132, where the error is returned! Thank you for precious time and I am applying the changes as you suggest!